File: moneymarket/src/main/java/com/example/moneymarket/service/InterestRateService.java
package com.example.moneymarket.service;

import com.example.moneymarket.dto.InterestRateResponseDTO;
import com.example.moneymarket.entity.InterestRateMaster;
import com.example.moneymarket.exception.ResourceNotFoundException;
import com.example.moneymarket.repository.InterestRateMasterRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class InterestRateService {

    private final InterestRateMasterRepository interestRateMasterRepository;

    public List<InterestRateResponseDTO> getAllInterestRates() {
        List<InterestRateMaster> interestRates = interestRateMasterRepository.findAll();
        return interestRates.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public InterestRateResponseDTO getInterestRateByCode(String inttCode) {
        InterestRateMaster interestRate = interestRateMasterRepository.findById(inttCode)
                .orElseThrow(() -> new ResourceNotFoundException("Interest Rate", "Code", inttCode));
        return mapToResponse(interestRate);
    }

    private InterestRateResponseDTO mapToResponse(InterestRateMaster entity) {
        return InterestRateResponseDTO.builder()
                .inttCode(entity.getInttCode())
                .inttRate(entity.getInttRate())
                .inttEffctvDate(entity.getInttEffctvDate())
                .build();
    }
}

File: frontend/src/api/interestRateService.ts
import apiClient from './apiClient';

export interface InterestRateMaster {
  inttCode: string;
  inttRate: number;
  inttEffctvDate: string;
}

const INTEREST_RATE_ENDPOINT = '/interest-rates';

export const getAllInterestRates = async (): Promise<InterestRateMaster[]> => {
  const response = await apiClient.get(INTEREST_RATE_ENDPOINT);
  return response.data;
};

export const getInterestRateByCode = async (inttCode: string): Promise<InterestRateMaster> => {
  const response = await apiClient.get(`${INTEREST_RATE_ENDPOINT}/${inttCode}`);
  return response.data;
};
